adabot/lib/circuitpython_library_validators.py

```diff
@@ -372,41 +371,71 @@ def _validate_py_for_u_modules(self, repo, download_url):
+                - Pylint version is the latest release
         """
-        return []
 
-        download_url = travis_yml_file_info["download_url"]
+        download_url = actions_build_info["download_url"]
         contents = requests.get(download_url, timeout=30)
         if not contents.ok:
             return [ERROR_PYFILE_DOWNLOAD_FAILED]
 
         errors = []
 
-        lines = contents.text.split("\n")
-        pypi_providers_lines = (
-            [l for l in lines
-            if re.match(r"[\s]*-[\s]*provider:[\s]*pypi[\s]*", l)]
+        pylint_version = None
+        re_pip_pattern = r"pip\sinstall.*"
+        re_pylint_pattern = (
+            r"pylint(?P<eval>(?:[<>~=]){0,2})(?P<major>\d*)(?P<minor>(?:\.\d){0,2})"
         )
 
-        if not pypi_providers_lines:
-            errors.append(ERROR_MISSING_PYPIPROVIDER)
+        pip_line = re.search(re_pip_pattern, contents.text)
+        if not pip_line:
+            return [ERROR_PYLINT_VERSION_NOT_FIXED]
+
+        pip_line = pip_line[0]
+
+        pylint_info = re.search(re_pylint_pattern, pip_line)
+        if not pylint_info:
+            return [ERROR_PYLINT_VERSION_NOT_FIXED]
+
+        if pylint_info.group("eval"):
+            pylint_version = (
+                f"{pylint_info.group('major')}"
+                f"{pylint_info.group('minor').replace('.', '')}"
+            )
+            eval_func = pylint_info.group("eval")
+            eval_len = len(pylint_version)
+
+            if "<" in eval_func:
+                eval_str = (
+                    f"{self.latest_pylint[:eval_len]} "
+                    f"{eval_func} "
+                    f"{pylint_version}"
+                )
+            elif "~" not in eval_func:
+                eval_str = (
+                    f"{pylint_version} "
+                    f"{eval_func} "
+                    f"{self.latest_pylint[:eval_len]}"
+                )
+            else:
+                return [ERROR_PYLINT_VERSION_NOT_FIXED]
 
-        pylint_version = None
-        for line in lines:
-            if not line.strip().startswith("- pip install --force-reinstall pylint=="):
-                continue
-            pylint_version = line.split("=")[-1]
+        else:
+            pylint_version = self.latest_pylint
+            eval_str = "True"
+
+        #print(
+        #    f"{repo['name']}: pylint_version: {pylint_version} latest: {self.latest_pylint}\n"
+        #    f"{pylint_info.groups()} eval_str: {eval_str}"
+        #)
 
         if not pylint_version:
             errors.append(ERROR_PYLINT_VERSION_NOT_FIXED)
-        # disabling below for now, since we know all pylint versions are old
-        # will re-enable once efforts are underway to update pylint
-        #elif pylint_version.startswith("1."):
-        #    errors.append(ERROR_PYLINT_VERSION_VERY_OUTDATED)
-        #elif pylint_version != self.latest_pylint:
-        #    errors.append(ERROR_PYLINT_VERSION_NOT_LATEST)
+        elif not eval(eval_str):
```

Comment:
Calling eval scares me. Is there a python package that can be used for analysis of the version string? (I bet there is.)

Reply:
I understand the reservations on `eval()`. I have them too. This is why I limit the eval to using the `eval_str` that is built internally, based on the regex results.

I'll revisit it. This was the expedient solution, since the effort is already underway...